examples/cpp/Schrodinger2D.cpp

```diff
@@ -0,0 +1,156 @@
+#include <iomanip>    
+#include <vector>    
+#include <fstream>    
+  
+using namespace arma;    
+  
+int main() {    
+  int p = 105;  // Number of time steps for the simulation    
+  double Lxy = 1.0;    
+  int k = 2;  // Order of accuracy    
+  int m = 50; // Grid points in x    
+  int n = 50; // Grid points in y    
+  int nx = 2; // Energy level in x    
+  int ny = 2; // Energy level in y    
+  double dx = Lxy / m; // Step in x    
+  double dy = Lxy / n; // Step in y    
+  double dt = dx/2; // Reduced time step size with additional reduction    
+  
+  // Define staggered grids    
+  vec xgrid = join_vert(vec({0}), linspace(dx / 2, Lxy - dx / 2, m), vec({Lxy}));    
+  vec ygrid = join_vert(vec({0}), linspace(dy / 2, Lxy - dy / 2, n), vec({Lxy}));    
+  
+  // Initialize 2D staggered grid    
+  mat X, Y;    
+  Utils utils;    
+  utils.meshgrid(xgrid, ygrid, X, Y);    
+  
+  // Debug grid sizes    
+  std::cout << "Grid X size: " << X.n_rows << " x " << X.n_cols << std::endl;    
+  std::cout << "Grid Y size: " << Y.n_rows << " x " << Y.n_cols << std::endl;    
+  
+  // Initialize Laplacian operator with Robin BC    
+  Laplacian L(k, m, n, dx, dy);    
+  RobinBC BC(k, m, 1, n, 1, 1, 0);    
+  L = L + BC;    
+  
+  // Ensure the Laplacian is square    
+  int total_size = (m + 2) * (n + 2); // Total size for the grid including boundaries    
+  std::cout << "Laplacian L has size: " << L.n_rows << " x " << L.n_cols << std::endl;    
+  
+  // Hamiltonian definition with damping term    
+  auto H = [&](const vec &x) {    
+   vec result = 0.5 * (L * x);    
+   for (size_t i = 0; i < x.n_elem; ++i) {    
+    result(i) -= 0.01 * x(i); // Damping term    
+   }    
+   return result;    
+  };    
+  
+  // Define wave numbers    
+  auto kx = [&](int nx) { return nx * M_PI / Lxy; };    
+  auto ky = [&](int ny) { return ny * M_PI / Lxy; };    
+  
+  double A = 2 / Lxy;    
+  
+  // Initialize the wavefunction psi_old    
+  mat Psi_grid(m + 2, n + 2, fill::zeros);    
+  
+  // Manually pad Psi_grid    
+  for (int i = 1; i <= m; i++) {    
+   for (int j = 1; j <= n; j++) {    
+    Psi_grid(i, j) = A * sin(kx(nx) * X(i, j)) * sin(ky(ny) * Y(i, j));    
+   }    
+  }    
+  
+  // Convert to column vector for compatibility    
+  vec psi_old = vectorise(Psi_grid);  // Convert to column vector    
+  
+  // Debug size of psi_old    
+  std::cout << "psi_old size: " << psi_old.n_elem << ", expected: " << total_size << std::endl;    
+  
+  if (psi_old.n_elem != total_size) {    
+   std::cerr << "Error: psi_old size mismatch. Expected " << total_size    
+    << ", got " << psi_old.n_elem << std::endl;    
+   return -1;    
+  }    
+  
+  vec v_old(total_size, fill::zeros); // Initialize v_old to the same size as psi_old    
+  
+  // Initialize Psi_re with zeros    
+  mat Psi_re = zeros<mat>(m + 2, n + 2);    
+  
+  // Create 2D interpolator of the second type    
+  Interpol I2(true, m, n, 0.5, 0.5);    

```

Comment:
This is equivalent to the interpolD2D in MATLAB. You also need the interpol2D.
use the same call but omit the 'true'
```  Interpol I(m, n, 0.5, 0.5);    ```